[addons] move binary addon interface functions to own files / cleanups

AlwinEsch opened the pull request:
## Description
This includes a change which brings the binary addon interface functions (addon to Kodi) into own files. This is intended to clean up the interface.

In addition, the primary task of the CAddonDll is to load the addon for Kodi and make it available, but not to provide its interface itself.

The interface folder therefore refers purely to the API between addon and Kodi, therefore all API functions should also be in it if possible (partly except for Kodi to addon instances).

Furthermore, it contains some code cleanups, so please review the commits individually for review.

The commits from https://github.com/xbmc/xbmc/pull/17292 are temporarily there (was to avoid conflicts) which disappear when it is in there.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
To make the addon interface cleaner
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] add some missing 'extern "C"' to binary interface calls

AlwinEsch committed:
[addons] remove wrong included header "FileItem.h" from Interface_Filesystem

On the interface headers used by the addon, Kodi's own headers should be
avoided as much as possible! This avoids inadvertently used things that do
not correspond to the API.

AlwinEsch committed:
[addons] some clang cleanups on binary addon interface headers

AlwinEsch committed:
[addons] change few CAddonDll parts to match more C++ style

phunkyfish reviewed (approved):
Nice cleanup.

garbear commented on xbmc/addons/interfaces/AddonBase.cpp:
```diff
@@ -192,39 +192,39 @@ void Interface_Base::addon_log_msg(void* kodiBase, const int addonLogLevel, cons
   if (addon == nullptr || id == nullptr || value == nullptr)
   {
-    CLog::Log(LOGERROR, "Interface_Base::%s - invalid data (addon='%p', id='%p', value='%p')",
+    CLog::Log(LOGERROR, "Interface_Base::{} - invalid data (addon='%p', id='%p', value='%p')",
```
Why does `%p` stay old style?

AlwinEsch replied:
Was not 100% sure it makes in this case sense to see better it goes about a pointer, seen also some parts now where not use the `static_cast<void*>` for them an want to change.

AlwinEsch replied:
Seen during test it is not only a can be with `{}`, it is a must be then.
This was given on test:
```
ERROR: Interface_Base::get_setting_bool - invalid data (addon='%p', id='%p', value='%p')
```
Only the `get_setting_bool` was correct the rest has written `%p` to log.

AlwinEsch committed:
[addons] change some binary interface log calls to new style

AlwinEsch committed:
[addons] rename binary gui interface dir to "gui" (lowercase) to match headers

AlwinEsch committed:
[addons] code style cleanup on Interface_Base to match more C++

With C++ only for inside Kodi controlled parts.

AlwinEsch committed:
[addons] move binary addon interface calls to own files

Previously, these functions were within CAddonDll, but since its function
is primarily the basis for Kodi of the addon, but not its interface, these
functions did not fit there.

Since these form the API, they belong in the associated interface folder.
This allows a much better overview, and the name "AddonBase.*"
used there corresponds to the associated addon header.
